Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,294 @@
+        requires left(pos).Success? ==> right.requires(left(pos).pos)
+      =>
+        var (pos, l) :- left(pos);
+        var (pos, r) :- right(pos);
+        Success(pos, (l, r))
+    }
+
+/// The two important things to note are (1) the fact that `Concat` is now
+/// marked `:opaque`, and (2) the fact that `Concat` now has an `ensures` clause
+/// of the following form:
+///
+/// ```dafny
+///   ensures forall input :: (precondition of the parser) ==> p.requires(input)
+/// ```
+///
+/// This does not say that the parser must accept all inputs: only that the
+/// condition `(precondition of the parser)` is sufficient to call the opaque
+/// function `p` that `Concat` returns.
+///
+/// With that in mind, we can define `Either` in the same way:
+
+    static function {:opaque} Either<L, R>(
+      left: Parser<L>,
+      right: Parser<R>
+    )
+      : (p: Parser<Either<L, R>>)
+      ensures forall pos: nat |
+        && left.requires(pos)
+        && (left(pos).Failure? ==> right.requires(pos))
+        :: p.requires(pos)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Failure? ==> right.requires(pos)
+      =>
+        match left(pos)
+          case Failure => right(pos).MapResult(r => Right(r))
+          case Success(pos, l) => Success(pos, Left(l))
+    }
+
+/// Next, we define `Char`, which checks for a specific character and advances
+/// the cursor by one.  Its post-condition indicates that the cursor has
+/// advanced.  It additionally states that the new position is `<= |input|`
+/// (which is equivalent to saying that the original position was `< |input|`).
+
+    function {:opaque} Char(c: char): (p: Parser<char>)
+      ensures forall pos: nat :: p.requires(pos)
+      ensures forall pos: nat :: p(pos).Success? ==>
+        p(pos).pos == pos + 1 && p(pos).pos <= |input|
+    {
+      (pos: nat) =>
+        if pos < |input| && input[pos] == c then Success(pos + 1, c)
+        else Failure()
+    }
+
+/// Finally, `EOS` (“End Of String”) checks whether all input has been consumed…
+
+    const EOS: Parser<()> :=
+      (pos: nat) => if pos >= |input| then Success(pos, ()) else Failure()
+
+/// … and `Epsilon` matches the empty string:
+
+    static const Epsilon: Parser<()> :=
+      (pos: nat) => Success(pos, ())
+
+/// These combinators are enough to define a parser for parentheses.  Note that
+/// because the parser calls itself, we need to eta-expand it (so that we can
+/// mention `pos` in its spec).  Also note the call to `MapResult`, which
+/// implements a simple semantic action (here, counting parentheses).
+
+    function {:opaque} Parentheses'(pos: nat): ParseResult<nat>
+      decreases |input| - pos
+    {
+      var rec := pos' requires pos < pos' <= |input| => Parentheses'(pos');
+      Either(Concat(Char('('), Concat(rec, Char(')'))), Epsilon)(pos)
+        .MapResult<nat>(
+          var count := (r: Either<(char, (nat, char)), ()>) =>
+            match r
```

Comment:
I like extra braces for the match. Even if they're not necessary, I think it is good practice for nested matches.

Reply:
I agree for nested matches, but isn't it verbose for non-nested matches?

Reply:
I prefer without parentheses here.